<a href="https://colab.research.google.com/github/EduardoAlvesP/ml-study/blob/main/01_erro_regressao.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

AULA 1 INTRODUÇÃO

In [23]:
#Vamos supor que o valor real de uma linha é a soma de todos seus dados * 10
tetas = [4,6,2,10,8,4,20]
def preve(tetas,dados):
  a = 0
  for i in range(len(tetas)):
    a = a + tetas[i] * dados[i]
  return a

def valor_real(dados):
  a = 0
  for i in dados:
    a = a + i
  a = a * 10
  return a
def calcular_erro(linha,tetas):
  valor_calculado = preve(tetas,linha)
  valor = valor_real(linha)
  return pow(valor - valor_calculado,2)


dados = [
    [1, 12, 12, 0, 0, 0, 0],
    [1, 10,  9, 1, 0, 1, 2],
    [1,  8,  8, 1, 1, 2, 1],
    [1,  6,  9, 0, 2, 2, 0],
    [1,  9,  5, 2, 0, 0, 3],
    [1,  4,  7, 1, 2, 3, 1],
    [1,  7,  3, 3, 1, 1, 2],
]

def erros_matriz(dados,tetas):
  erro_total = 0
  for i in dados:
    erro_total = erro_total + calcular_erro(i,tetas)
  return erro_total
tetas1 = [4,6,2,10,8,4,20]
print(erros_matriz(dados,tetas1))
#Diminuindo o 20
tetas2 = [4,6,2,10,8,4,15]

print(erros_matriz(dados,tetas2))
#deixando tudo >= 5
tetas3 = [5,6,5,10,8,5,15]
print(erros_matriz(dados,tetas3))
# deixando alguns com peso 10
tetas4 = [5,10,5,10,10,5,10]
print(erros_matriz(dados,tetas4))
# deixando todos com peso 10
tetas5 = [10,10,10,10,10,10,10]
print(erros_matriz(dados,tetas5))

71396
78391
44320
18425
0


## O que eu aprendi

**Modelo:** o chute é uma soma: cada dado vezes o seu peso, tudo somado. `chute = θ0·x0 + θ1·x1 + ... + θ6·x6`
- **x** são os dados de cada linha, mudam em cada uma.
- **θ** são os pesos, os mesmos pra todas as linhas. É isso que o modelo aprende.
- **x0 é sempre 1**, pra o θ0 funcionar como uma constante (bias). O θ0 em si não é 1, quem é 1 é o x0.

**Função target:** é o valor real, o certo. Aqui eu inventei que é a soma dos dados vezes 10. Na vida real ninguém sabe essa função, só alguns exemplos dela.

**Erro:** pra cada linha faz `(real − chute)²` e depois soma todas as linhas. O quadrado é pra erro positivo e negativo não se cancelarem, e pra erro grande pesar mais.

Como o real é a soma × 10, os pesos certos são todos 10, e aí o erro dá 0.

### O que aconteceu no tetas2
Baixei o θ6 de 20 pra 15 e o erro **piorou** (71396 → 78391), mesmo 15 estando mais perto do certo, que é 10.
Isso aconteceu porque o modelo tava chutando **abaixo** do real em todas as linhas, então baixar qualquer θ abaixa o chute e piora.
Conclusão: o erro só vê a soma final. Eu tinha que subir a soma como um todo primeiro.

### Ideia do gradiente (próximo passo)
Pra cada θ, cada linha dá um "voto": `erro da linha × dado daquele θ na linha`.
- O erro diz pra que lado ir: positivo quer dizer que o chute ficou baixo, então tem que subir o θ.
- O dado diz a força: dado grande quer dizer que o θ mexe muito naquela linha, e dado 0 quer dizer que a linha não vota.

Soma os votos de todas as linhas e atualiza: `θ = θ + lr × voto`.
O `lr` é o mesmo pra todos os θ e só controla o tamanho do passo.
Repete isso várias vezes até o erro ficar perto de 0.